vllm/model_executor/layers/quantization/fp8.py

```diff
@@ -81,46 +98,163 @@ def create_weights(
+
+        return param[shard_id], loaded_weight
 
     def process_weights_after_loading(self, layer: Module) -> None:
-        # Although the quant_method is propagated to all layers,
-        # only linear layers invoke "create_weights". So we check
-        # whether "weight_scaling_facor" is registered to determine
-        # whether the layer is a linear layer that requires quantization.
-        if not hasattr(layer, "weight_scaling_factor"):
+        if (not hasattr(layer, "process_after_load")
+                or not layer.process_after_load):
             return
 
-        qweight, weight_scale = ops.scaled_fp8_quant(layer.weight)
-        # torch._scaled_mm requires column-major in the second
-        # input (weight), so we transpose the quantized weight.
-        layer.weight = Parameter(qweight.t(), requires_grad=False)
-        layer.weight_scaling_factor.data.copy_(weight_scale)
+        # If checkpoint is fp/bf16 (not serialized fp8), quantize the weights.
+        if not self.quant_config.is_serialized:
+            qweight, weight_scale = ops.scaled_fp8_quant(layer.weight,
+                                                         scale=None)
+            layer.weight = Parameter(qweight.t(), requires_grad=False)
+            layer.weight_scale = Parameter(weight_scale, requires_grad=False)
+            layer.logical_widths = None
+            layer.act_scale = None
+            return
+
+        # TODO: cutlass kernels will remove the need for much of this logic.
+        # If the checkpoint is serialized fp8, we already loaded quantized,
+        # so, just cleanup the Parameters for easier use in apply().
+        else:
+            # WEIGHT
+            #   Transpose weight for passing to torch._scaled_mm
+            weight = layer.weight
+            layer.weight = Parameter(weight.t(), requires_grad=False)
+
+            # WEIGHT_SCALE
+            #   If all weight_scales are equal, use a single scale.
+            if all_close_1d(layer.weight_scale):
+                layer.weight_scale = Parameter(layer.weight_scale.max(),
+                                               requires_grad=False)
+                layer.logical_widths = None
+
+            # ACT_SCALE
+            #   Dynamic: set to None (required input to ops.scaled_fp8_quant).
+            #   Static:  set to max of the act_scales (since they are equal).
+            if self.quant_config.activation_scheme == "dynamic":
+                layer.act_scale = None
+            elif self.quant_config.activation_scheme == "static":
+                if not all_close_1d(layer.act_scale):
+                    raise ValueError(
+                        "All the act_scales for the logical weights of a layer "
+                        f"must be equal. But got {layer.act_scale}")
+                layer.act_scale = Parameter(layer.act_scale.max(),
+                                            requires_grad=False)
```

Comment:
Ditto. Does this mean we enforce activation to be quantized per-tensor?

Reply:
This just checks that the `act_scale` for `q_proj`, `k_proj`, and `v_proj` are equal.

This will always be true, unless something is wrong with the checkpoint

Reply:
But once we support something else, we can relax this (though I dont really think anything besides per tensor makes sense for static activation quantization)